### PR TITLE
mpi: Move to preallocated request objects + collectives

### PR DESCRIPTION
**Pull request by AntonLydike**
This PR changes the MPI dialect to use pre-allocated request objects instead of allocating one per `I`-operation.

It introduces a special `VectorType` which wraps a number of `Request` or `Status` or `Datatype` objects. These are lowered to llvm pointers.

This also adds support for collectives.

### CODE REVIEW

**Comment by AntonLydike** [comment]
Note that this PR currently contains some of #591's commits

**Review by webmiche — COMMENTED** [review]
Would it be possible to split this PR? Maybe remove everything but the OperationType and an example op, say the `ReduceOp`? Or is this on the critical path for the paper? In that case, I would prefer to move this to the experimental folder, if you want to iterate more quickly.

**Comment by AntonLydike** [comment]
@webmiche is this smallenough to

**Comment by AntonLydike** [comment]
@webmiche is this small enough to review, or should I maybe separate out some more stuff?

File: tests/dialects/test_mpi_lowering.py [hunk 1]
```diff
@@ -322,6 +326,37 @@ def test_lower_mpi_bcast():
     )
 
 
+def test_lower_mpi_allocate():
+    count, = CreateTestValsOp.get(i32).results
+    op = mpi.AllocateTypeOp.get(mpi.RequestType, count)
+
+    ops, res = lower_mpi.LowerMpiAllocateType(info).lower(op)
+
+    assert len(res) == 1
```
Comment (math-fehr) [inline tests/dialects/test_mpi_lowering.py:335]: nit: Why are these lines so separated?

**Review by math-fehr — APPROVED** [review]
Looks good to me!

File: xdsl/dialects/mpi.py [hunk 2]
```diff
@@ -78,6 +80,23 @@ class DataType(ParametrizedAttribute, MLIRType):
     name = 'mpi.datatype'
 
 
+VectorWrappable = RequestType | StatusType | DataType
+_VectorT = TypeVar('_VectorT', bound=VectorWrappable)
+
+
+@irdl_attr_definition
+class VectorType(Generic[_VectorT], ParametrizedAttribute, MLIRType):
```
Comment (webmiche) [inline xdsl/dialects/mpi.py:88]: Why is the builtin vectortype not enough here?

File: xdsl/dialects/mpi.py [hunk 3]
```diff
@@ -214,22 +248,21 @@ class Bcast(MPIBaseOp):
...
+    @staticmethod
     def get(
-        cls,
         buffer: SSAValue | Operation,
         count: SSAValue | Operation,
         datatype: SSAValue | Operation,
         root: SSAValue | Operation,
     ):
-        return cls.build(
+        return Bcast.build(
```
Comment (webmiche) [inline xdsl/dialects/mpi.py:258]: :heart_eyes:

File: xdsl/dialects/mpi.py [hunk 4]
```diff
@@ -317,7 +350,7 @@ def get(buffer: SSAValue | Operation, count: SSAValue | Operation,
 
 
 @irdl_op_definition
-class IRecv(MPIBaseOp):
+class Irecv(MPIBaseOp):
```
Comment (webmiche) [inline xdsl/dialects/mpi.py:353]: I slightly prefer the two capitals at the start, but I don't mind too much.

**Review by webmiche — APPROVED** [review]
LGTM, modulo minor stuff. Also thanks for splitting, this was way easier this way :)

File: xdsl/dialects/mpi.py [hunk 5]
```diff
@@ -468,6 +501,38 @@ def get(request: Operand, ignore_status: bool = True):
         return Wait.build(operands=[request], result_types=result_types)
 
 
+@irdl_op_definition
+class WaitAll(MPIBaseOp):
```
Comment (superlopuh) [inline xdsl/dialects/mpi.py:505]: nit: if we're being consistent with the MPI names, then maybe this should be called `Waitall`
```suggestion
class Waitall(MPIBaseOp):
```

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/mpi.py (hunk 4 above)
Comment (georgebisbas) [reply xdsl/dialects/mpi.py:353]: Same

Comment (AntonLydike) [reply xdsl/dialects/mpi.py:353]: It's to be "consistent" with MPI naming. It's called "Irecv" there, sadly :(

File: xdsl/dialects/mpi.py (hunk 2 above)
Comment (AntonLydike) [reply xdsl/dialects/mpi.py:88]: I'm not 100% sure, I think I looked at the docs for `vector` and was put off by the fact that it was specifically about stuff that should be vectorized.

We can def. move that to the builtin `vector` later on!

**Pull request merged**